### DISCUSSION TITLE
Scripting/Smart Meter Interface: how to send a specific serial signal for a very long time

### DISCUSSION THREAD

**Discussion by christophhhh** [arendst/Tasmota#17388] [Q&A]
Hello,
this is related to topic [#17283](https://github.com/arendst/Tasmota/discussions/17283)

I want to get data from a rather complicated heat meter. For that I have to:

- **send 0x55 for 2,2 seconds with 8E1, 2400 baud (wakeup sequence)**
- wait 400 ms
- send 0x10 0x40 0xFE 0x3E 0x16 with 8N1, 2400 baud
- wait and receive the answer 

This is what I tried so far:
- use the `switch` selector in the part of the script that is executed every 100ms (`>F` section) to jump through the different states
- have a loop in the first state to send `0x55` 532 times which should take 2216ms (with 2400 baud)

So far this is not successfull. I could not find documentation about what happens if a loop in the `>F` section takes longer than 100ms. It seems like the script is not executed at all.

Thanks in advance!

Here is the entire code:
```
>D
;start, define variables
res=0
scnt=0
wkup=1


>B
;setup sensor
->sensor53 r

>F
;this section is executed every 100 ms
scnt+=1
;increment counter

switch scnt
case 1
for wkup 1 10 1
res=sml(1 1 "55")
next
;set string to send to "55" on meter 2 (8E1) 532 times (wake up sequence)
;this should take 2216ms (as required by meter)

res=sml(1 2)
;change to read mode, send nothing

case 26
;   0,1s (start delay)
; + 2,2s wakeup period 
; + 0,3s waiting time (as required by meter)
;is over, we can send the request for data

res=sml(2 1 "1040FE3E16")
;set string to send to "1040FE3E16" on meter 2 (8N1) (request data)

res=sml(2 2)
;set to read mode

case 100
;after 10s > restart sequence
scnt=0
ends

>M 2
;declare sensors

+1,3,M,0,2400,WAERME,1
;Sensor 1: 8E1

+2,3,m,0,2400,WAERME,1
;Sensor 2: 8N1
define variables
res=0
scnt=0
wkup=1

>B
;setup sensor
->sensor53 r

>F
;this section is executed every 100 ms
scnt+=1
;increment counter

switch scnt
case 1
for wkup 1 532 1
res=sml(1 1 "55")
next
;set string to send to "55" on meter 2 (8E1) 532 times (wake up sequence)
;this should take 2216ms (as required by meter)

res=sml(1 2)
;change to read mode, send nothing

case 26
; 0,1s (start delay)
; + 2,2s wakeup period
; + 0,3s waiting time (as required by meter)
;is over, we can send the request for data

res=sml(2 1 "1040FE3E16")
;set string to send to "1040FE3E16" on meter 2 (8N1) (request data)

res=sml(2 2)
;set to read mode

case 100
;after 10s > restart sequence
scnt=0
ends

>M 2
;declare sensors

+1,3,M,0,2400,WAERME,1
;Sensor 1: 8E1

+2,3,m,0,2400,WAERME,1
;Sensor 2: 8N1
```

**Comment by gemu2015** [comment]
defining 2 meters at the same port will not work at all.
you must use the new cmd i introduced in the other thread.

since this kind of meter can only be called several times a day i would do this in a subroutine that you call e.g. every 4 hours. in this case you do not need a state machine, simply use delay(x)
don't mind about stopping the main loop for about 3 seconds, it will not generate a watchdog event.

define 1 meter with 8N1

;set serial protocol
sml(-1 1 "2400:8E1")
;send 0x55 for 2,2 seconds with 8E1, 2400 baud (wakeup sequence)
for wkup 1 532 1
sml(1 1 "55")
next 
delay(400)
;switch serial protocol
sml(-1 1 "2400:8N1")
;set string to send to "1040FE3E16" on meter 2 (8N1) (request data)
sml(2 1 "1040FE3E16")

you should now receive your meters response

**Reply by christophhhh** [reply]
Hi Gemu,
thanks for your reply! I compiled and flashed a firmware with your addition. Below is my current script that executes the subroutine once after saving the script. 

I have the following questions:
- how do I manage to execute the subroutine #readmeter every 4h let's say?
- when enabling dump mode with `sensor53 d1` I would expect to see at least what I send (many "55", the "request data string"), but I only get the "print" output
- do I understand it correctly, that to switch the serial protocol I have to use this command: `sml(-1 1 "2400:8N1")` (**minus** 1 to reset serial protocol on the first sensor)

output: (i expect to see at least what I send..)
```
23:11:13.576 CMD: sensor53 d1
23:11:13.581 RSL: RESULT = {"Time":"2022-12-14T23:11:13","SML":{"CMD":"dump: 1"}}
23:11:46.067 script compressed to 384 bytes = 66 %
23:11:46.070 Script: nv=1, tv=0, vns=5, vmem=32, smem=2560
23:11:46.087 start
23:11:51.093 waited 5s, call subroutine
23:11:51.095 wake up meter
23:11:51.125 wait for the meter
23:11:51.527 request data
```

script:

```
>D
;start, define variables
wkup=1


>B
;setup sensor
->sensor53 r

>BS
print start
delay(5000)
print waited 5s, call subroutine

=#readmeter

#readmeter
print wake up meter
;set serial protocol
sml(-1 1 "2400:8E1")
;send 0x55 for 2,2 seconds with 8E1, 2400 baud (wakeup sequence)
for wkup 1 532 1
sml(1 1 "55")
next
wkup=1
print wait for the meter
;wait for the meter to wake up
delay(400)
;switch serial protocol
sml(-1 1 "2400:8N1")
print request data
;set string to send to "1040FE3E16" (request data)
sml(1 1 "1040FE3E16")
#


>M 1
+1,3,M,0,2400,WAERME,1
;Sensor 1: 8E1
#
```

**Reply by gemu2015** [reply]
```
>S
if upsecs%60==0 {
=#readmeter
}
```
e.g. calls the subroutine every 60 seconds

you won't see serial data sent with sml(1 1 ...) at the console

yes the new cmd resets the serial port when given a negativ meter number

btw you may not use the 'M' for this meter. 'M' means MODBUS not MBUS.

you must use 'r' for RAW decoder

here is an example of a working MBUS Engelmann Sensostar U script from another forum:

```
>D
>B
->sensor53 r
>M 1
+1,3,rE1,0,2400,Waerme,1,10,1040FD3D16,105B005B16


1,68b3b368x23uuUUUUUU@1,Energie total,,total,0 ; Total Wärmemenge
1,68b3b368x47uuUUUUUU@60,Flow F_akt,,F_akt,0 ; aktueller Flow in m^3
1,68b3b368x35uuUUUU@1,Power P_akt,,P_akt,0 ; aktuelle Wärme-Leistung
1,68b3b368x59uuUU@1,Temp Vorlauf,,Vorlauf,0 ; aktuelle Vorlauftemperatur
1,68b3b368x63uuUU@1,Temp Rückl,,Rückl,0 ; aktuelle Vorlauftemperatur
1,68b3b368x67uuUU@1,Temp Diff,,Diff,0 ; aktuelle Differenz Vorlauf/Rücklauftemperatur
#
```

remark:
this meter does not need a wakeup sequence


you also need to increase the serial buffer size because of larger responses from these meters

e.g.

#define SML_BSIZ 200

**Reply by christophhhh** [reply]
thanks for your reply. Unforunately I still can't get it running.
- where should I add `#define SML_BSIZ 200` ? 
- I found the definition in `xsns_sml_53.ino` and modified the buffer size there. Was that right or should I add something like
```
#ifndef SML_BSIZ
#define SML_BSIZ 200
#endif
``` 
to `user_config_override.h`?

I got an successfull reply with another software. Below I postet the answer the meter sends.
- is the buffer size 200 big enough?
- is there any documentation on how to write decoders for this data?


```
Port Modus 8N1
wakeup(): Sende Aufwachsequenz
Pause 350 ms
Port Modus 8E1
req_ud2(): Sende Anfrage Daten Klasse 2
[10][5B][FE][59][16]
5 bytes sent
mbus_bytes_received: 8
[68] [4D] [4D] [68] [08] [00] [72] [84]
mbus_bytes_received: 8
[03] [36] [20] [92] [26] [17] [04] [01]
mbus_bytes_received: 8
[00] [00] [00] [0C] [78] [84] [03] [36]
mbus_bytes_received: 8
[20] [04] [06] [54] [61] [00] [00] [0C]
mbus_bytes_received: 8
[14] [46] [57] [08] [00] [0B] [2D] [00]
mbus_bytes_received: 8
[00] [00] [0B] [3B] [00] [00] [00] [0A]
mbus_bytes_received: 8
[5A] [89] [04] [0A] [5E] [28] [04] [0B]
mbus_bytes_received: 8
[61] [03] [06] [00] [04] [6D] [36] [07]
mbus_bytes_received: 8
[C5] [2C] [02] [27] [1F] [04] [09] [FD]
mbus_bytes_received: 8
[0E] [07] [09] [FD] [0F] [11] [0F] [00]
mbus_bytes_received: 3
[00] [C3] [16]
83 byte(s) received
Data:
{0x68, 0x4D, 0x4D, 0x68, 0x08, 0x00, 0x72, 0x84, 0x03, 0x36, 0x20, 0x92, 0x26, 0x17, 0x04, 0x01, 0x00, 0x00, 0x00, 0x0C, 0x78, 0x84, 0x03, 0x36, 0x20, 0x04, 0x06, 0x54, 0x61, 0x00, 0x00, 0x0C, 0x14, 0x46, 0x57, 0x08, 0x00, 0x0B, 0x2D, 0x00, 0x00, 0x00, 0x0B, 0x3B, 0x00, 0x00, 0x00, 0x0A, 0x5A, 0x89, 0x04, 0x0A, 0x5E, 0x28, 0x04, 0x0B, 0x61, 0x03, 0x06, 0x00, 0x04, 0x6D, 0x36, 0x07, 0xC5, 0x2C, 0x02, 0x27, 0x1F, 0x04, 0x09, 0xFD, 0x0E, 0x07, 0x09, 0xFD, 0x0F, 0x11, 0x0F, 0x00, 0x00, 0xC3, 0x16, }
Length:         77
Checksum OK
C-Field:                0x08
Primary Adress:         0x00
Ci-Field:               0x72
variable Antwort, fixed header 12 Bytes

DIF 0C
VIF 78
        function_field: 00
        data_length:    4
        data_field:     0C
        unit:           78
        data bytes: 84 03 36 20
        Instantaneous value
        Fabrication number
        mbus_parse_data_field: DIF 0x0C was decoded using 8 digit BCD
        result: 20360384

DIF 04
VIF 06
        function_field: 00
        data_length:    4
        data_field:     04
        unit:           06
        data bytes: 54 61 00 00
        Instantaneous value
        Energy (kWh)
        mbus_parse_data_field: DIF 0x04 was decoded using 4 byte integer
        result: 24916   factor 1.000000 n: 3    neg 0

DIF 0C
VIF 14
        function_field: 00
        data_length:    4
        data_field:     0C
        unit:           14
        data bytes: 46 57 08 00
        Instantaneous value
        Volume (1e-2  m^3)
        mbus_parse_data_field: DIF 0x0C was decoded using 8 digit BCD
        result: 85746   factor 0.010000

DIF 0B
VIF 2D
        function_field: 00
        data_length:    3
        data_field:     0B
        unit:           2D
        data bytes: 00 00 00
        Instantaneous value
        Power (100 W)
        mbus_parse_data_field: DIF 0x0B was decoded using 6 digit BCD
        result: 0       factor 100.000000

DIF 0B
VIF 3B
        function_field: 00
        data_length:    3
        data_field:     0B
        unit:           3B
        data bytes: 00 00 00
        Instantaneous value
        Volume flow (m m^3/h)
        mbus_parse_data_field: DIF 0x0B was decoded using 6 digit BCD
        result: 0       factor 0.001000

DIF 0A
VIF 5A
        function_field: 00
        data_length:    2
        data_field:     0A
        unit:           5A
        data bytes: 89 04
        Instantaneous value
        Flow temperature (1e-1 deg C)
        mbus_parse_data_field: DIF 0x0A was decoded using 4 digit BCD
        result: 489     factor 0.100000

DIF 0A
VIF 5E
        function_field: 00
        data_length:    2
        data_field:     0A
        unit:           5E
        data bytes: 28 04
        Instantaneous value
        Return temperature (1e-1 deg C)
        mbus_parse_data_field: DIF 0x0A was decoded using 4 digit BCD
        result: 428     factor 0.100000

DIF 0B
VIF 61
        function_field: 00
        data_length:    3
        data_field:     0B
        unit:           61
        data bytes: 03 06 00
        Instantaneous value
        Temperature Difference (1e-2  deg C)
        mbus_parse_data_field: DIF 0x0B was decoded using 6 digit BCD
        result: 603     factor 0.010000

DIF 04
VIF 6D
        function_field: 00
        data_length:    4
        data_field:     04
        unit:           6D
        data bytes: 36 07 C5 2C
        Instantaneous value
        Time Point (time & date)

DIF 02
VIF 27
        function_field: 00
        data_length:    2
        data_field:     02
        unit:           27
        data bytes: 1F 04
        Instantaneous value
        Operating time (days)
        mbus_parse_data_field: DIF 0x02 with 2 byte Time Point (date)
        result: 1055

DIF 09
VIF FD
VIFE 0E
        function_field: 00
        data_length:    1
        data_field:     09
        unit:           7D
        data bytes: 07
        Instantaneous value
        VIF special char

DIF 09
VIF FD
VIFE 0F
        function_field: 00
        data_length:    1
        data_field:     09
        unit:           7D
        data bytes: 11
        Instantaneous value
        VIF special char

DIF 0F
MANUFACTURER_SPECIFIC:
00 00
-------------------------------------------------
primary_address:        0x00
customer_number:        20360384
manufacturer:           ITR
generation:             23
medium:                 4        = Wärme
reading_counter:        1
error_code:             0
signature:              0
fabrication_number:     20360384
energy:                 24916 kWh
volume:                 857460 l
power:                  0 W
flow:                   0 l/h
supply_temp:            48.9 °C
return_temp:            42.8 °C
temp_difference:        6.03 K
date:                   05.12.2022
time:                   07:54:00
operating_time:         1055 d
firmware_version:       7
software_version:       17
alarm_code:             0
-------------------------------------------------```

**Reply by gemu2015** [reply]
normally you override #defines in user_config_override.h, but does not matter for testing
your meter seems to send 83 bytes, so 200 will be OK.

please post the source code that works and generates the output from above, and your current Tasmota script

**Comment by christophhhh** [comment]
Hi gemu2015, thanks for your Help.

I got this response using the software linked here [https://www.mikrocontroller.net/topic/438972#6103099](url)
It's run on a raspberry pi not the ESP.

I send 53x "55555555555555555555" as I got random restarts when sending 532x "55":
17:09:05.052 RSL: INFO3 = {"Info3":{"RestartReason":{"Exception":4,"Reason":"Software Watchdog","EPC":["4000dd38","00000000","00000000"],"EXCVADDR":"00000000","DEPC":"00000000"},"BootCount":31}}

This is my current code:
```
>D
;start, define variables
wkup=1


>B
;setup sensor
->sensor53 r

;>BS
;print start
;delay(5000)
;print waited 5s, call subroutine

>F
if upsecs%60==0 {
print read meter
=#readmeter
}
#

#readmeter

print wakeup start
;set serial protocol
sml(-1 1 "2400:8E1")

;send 0x55 for 2,2 seconds with 8E1 (53x), 2400 baud (wakeup sequence)
for wkup 1 53 1
sml(1 1 "55555555555555555555")
next
print wakeup end
wkup=1

print wait for the meter
;wait for the meter to wake up
delay(350)

;switch serial protocol
sml(-1 1 "2400:8N1")
print request data
;set string to send to "105BFE5916" (request data)
sml(1 1 "105BFE5916")



>M 1
+1,3,rE1,0,2400,WAERME,1
;Sensor 1: 8E1
;1,68b3b368x23uuUUUUUU@1,Energie total,,total,0 ; Total Wärmemenge
#
```

**Reply by gemu2015** [reply]
ok, you must revert serial definitions
the wakeup sequence is done with 8N1, the start code and receive must be with 8E1
and you should not use section >F (fast is executed every 100ms) but >S (executes every second)

**Reply by christophhhh** [reply]
Gemu, you made my day! After reverting the serial definitions I finally got an answer! I tried that in so many ways that apparently I messed it up in the end.
Here is the answer I get:
```
19:26:08.372 read meter
19:26:08.374 wakeup start
19:26:10.552 wakeup end
19:26:10.554 wait for the meter
19:26:10.907 request data
19:26:11.178 : 68 4d 4d 68 08 00 72 84 03 36 20 92 26 17 04 09 00 00 00 0c 78 84 03 36 20 04 06 aa 63 00 00 0c 14 69 75 08 00 0b 2d 00 
19:26:11.224 : 00 00 0b 3b 00 00 00 0a 5a 09 
19:26:11.268 : 05 0a 5e 95 02 0b 61 41 21 00 
19:26:11.313 : 04 6d 11 14 d2 2c 02 27 2c 
19:26:11.356 : 04 09 fd 0e 07 09 fd 0f 11 0f 
19:26:11.400 : 00 00 ab 16 
```

now I just have to decode this answer...

And here is my code:
```
>D
;start, define variables
wkup=1


>B
;setup sensor
->sensor53 r

;>BS
;print start
;delay(5000)
;print waited 5s, call subroutine

>S
if upsecs%60==0 {
print read meter
=#readmeter
}
#

#readmeter

print wakeup start
;set serial protocol
sml(-1 1 "2400:8N1")

;send 0x55 for 2,2 seconds with 8N1 (53x), 2400 baud (wakeup sequence)
for wkup 1 53 1
sml(1 1 "55555555555555555555")
next
print wakeup end
wkup=1

print wait for the meter
;wait for the meter to wake up
delay(350)

;switch serial protocol
sml(-1 1 "2400:8E1")
print request data
;set string to send to "105BFE5916" (request data)
sml(1 1 "105BFE5916")

>M 1
+1,3,rE1,0,2400,WAERME,1
;Sensor 1: 8E1
;1,68b3b368x23uuUUUUUU@1,Energie total,,total,0 ; Total Wärmemenge
#
```

**Reply by christophhhh** [reply]
This script will wake up and decode the data properly (apart from the timestamp).

This is the code used to decode the timestamp in the sample program I mentioned above:
```
else if (t_data_size == 4)           // Type F = Compound CP32: Date and Time
{
	if ((t_data[0] & 0x80) == 0)     // Time valid ?
	{
		t->tm_min = t_data[0] & 0x3F;
		t->tm_hour = t_data[1] & 0x1F;
		t->tm_mday = t_data[2] & 0x1F;
		t->tm_mon = (t_data[3] & 0x0F) - 1;
		t->tm_year = 100 + (((t_data[2] & 0xE0) >> 5) | ((t_data[3] & 0xF0) >> 1));
		t->tm_isdst = (t_data[1] & 0x80) ? 1 : 0;  // day saving time
	}
}
else
```

The SML documentation does not mention that the SML extension is able to decode bcd4/6/8 as well. Can I contribute to fix this?

Here is my code (timestamp is still wrong)
```
>D
;start, define variables
wkup=1


>B
;setup sensor
->sensor53 r

>S
if upsecs%3600==0 {
print read meter
=#readmeter
}
#

#readmeter

print wakeup start
;set serial protocol
sml(-1 1 "2400:8N1")

;send 0x55 for 2,2 seconds with 8N1 (53x), 2400 baud (wakeup sequence)
for wkup 1 53 1
sml(1 1 "55555555555555555555")
next
print wakeup end
wkup=1

print wait for the meter
;wait for the meter to wake up
delay(350)

;switch serial protocol
sml(-1 1 "2400:8E1")
print request data
;set string to send to "105BFE5916" (request data)
sml(1 1 "105BFE5916")



>M 1
+1,3,rE1,0,2400,WAERME,1
;Sensor 1: 8E1
1,0406uuUUUUUU@1,Total energy,kWh,w_total,0
1,0C14bcd8@100,Total volume,m³,v_total,2

1,0B2Dbcd6@0.01,Current power,W,p_act,0
1,0B3Bbcd6@1000,Current flow,m³/h,F_akt,3

1,0A5Abcd4@10,Flow temp,°C,t_flow,1
1,0A5Ebcd4@10,Return temp,°C,t_return,1
1,0B61bcd6@100,Temp diff,°C,t_diff,2

1,046Dbcd8@1,Time,,time,0
1,0227uuUU@1,Meter days,d,OpDays,0 
#
```

**Reply by gemu2015** [reply]
glad to see that you have success now.
ups, the bcd decoder is missing in the documentation.  would be nice if you add this and also your meter example to the docs.
the timestamp is very special and currently not supported.
i will think about a string decode option, or you may also decompose the bcd number derived from the timestamp with scripting

**Comment by carsten12** [comment]
Hi, first thanks to @gemu2015 for implementing the parity change. 

I have a similar device which I wanted to read with the same script as @christophhhh used. In fact we are discussing that in the same German forum and I followed how he managed. I am also using the WiFi head as him. I am sure the head works as it will work over WiFi with another power meter and I am also able to use the a.m. mbus-test sofware with an TTL/USB adapter on another computer. So I was even able to decode the M-BUS messages. For the records: the counter is an Elster 96 Plus which is identical to the Sharky 775.

I will try to describe my problem. I hope I don't make it too complicated, but I want you to have all information. In the end I kindly ask for your support how to further debug.

Here my script, after that I describe the small differences:

```
>D
;start, define variables
wkup=1


>B
;setup sensor
->sensor53 r

>S
if upsecs%60==0 {
print read meter
=#readmeter
}
#

#readmeter

print wakeup start
;set serial protocol
sml(-1 1 "2400:8N1")

;send 0x55 for 3 seconds with 8N1 (72x), 2400 baud (wakeup sequence)
for wkup 1 72 1
sml(1 1 "55555555555555555555")
next
print wakeup end
wkup=1

print wait for the meter
;wait for the meter to wake up
delay(350)

;switch serial protocol
sml(-1 1 "2400:8E1")
print request data
;set string to send to "105BFE5916" (request data)
sml(1 1 "6804046853FE5000A116"
sml(1 1 "105BFE5916")

>M 1
+1,3,rE1,0,2400,WAERME,1
1,0C06bcd8@1,Total Energy,kWh,w_total,0
1,0C13bcd8@1000,Total volume,m³,v_total,2
1,0C2Bbcd8@1,Current power,W,p_act,0
1,0B3Bbcd6@1000,Current flow,m³/h,F_akt,3
1,0A5Abcd4@10,Flow temp,°C,t_flow,1
1,0A5Ebcd4@10,Return temp,°C,t_return,1
1,0A62bcd4@10,Temp diff,°C,t_diff,2
#
```

Differences:
  - I use 72 times 0x55555555555555555555 because it is the mbus-test default
  - I also tried with 53x (original script) and 48x (meter spec says something about 480x 0x55)
  - I use SND_UD 0 message to reset what the meter will report, like in mbus-test but tried out without as well
  - Obviously my decoding templates are a little bit different    

I used the firmware of @christophhhh and I also built my own 12.3.1 with the following add-ons in user_config_override.h:

```
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif
#ifndef SML_BSIZ
#define SML_BSIZ 200
#endif
```

Unfortunately, the meter seems not to wake up. I can see the LED glow for the period of time expected, but nothing comes back. The head works, as when I don't place the head correctly, I will see the commands coming back as reflection. Also, I double checked that with mbus-test it works and head position is not problematic, as there is some mechanical border on the meter to show the necessary position.

But oposed to @christophhhh the logs don't show a period of 2-3 seconds:

```
18:15:28.392 read meter
18:15:28.394 wakeup start
18:15:28.400 wakeup end
18:15:28.401 wait for the meter
18:15:28.754 request data
```

Also, with mbus-test it will wait for an 0xF5 ACK message. 

I have two questions: 

1. Would I be able to wait (Inch timeout) for an 0xF5 although I am pretty sure it will not arrive, as I should see it when I type in sensor53 d1 in the console, should't I?
2. Do you have an idea how to make sure the wakeup sequence was really sent correctly? Is there any chance to debug something and see whether something has been sent?
3. Is there a way to visualise that the setting of baud rate and parity has been successfully done?

Many Thanks in advance and have a nice season!
Carsten.

**Reply by gemu2015** [reply]
you forgot 
#define **SML_SCRIPT_CMD**
try again and report back

**Reply by carsten12** [reply]
Unfortunately no change :-(

**Reply by gemu2015** [reply]
sorry typo => USE_SML_SCRIPT_CMD

**Reply by carsten12** [reply]
Ah, that solves one issue:

```
20:20:31.506 read meter
20:20:31.508 wakeup start
20:20:34.804 wakeup end
20:20:34.806 wait for the meter
20:20:35.159 request data
```

The timing now seems plausible. Nevertheless, nothing to be seen as sensor53 d1 output
so the meter is either now quiet as I have tried too often and did not recognise, or something
else is wrong. 

Do you have a suggestion how to debug?

Thanks,
Carsten.

**Reply by carsten12** [reply]
Another step further, when I skip the SND_UD 0 string I will receive a correct answer:

```
22:00:16.959 read meter
22:00:16.961 wakeup start
22:00:20.255 wakeup end
22:00:20.257 wait for the meter
22:00:20.609 request data
22:00:20.769 : 68 94 94 68 08 
22:00:20.814 : 00 72 02 13 49 44 93 15 2f 
22:00:20.858 : 04 73 00 00 00 0c 06 04 02 04 
22:00:20.901 : 00 8c 10 06 00 00 00 00 8c 
22:00:20.944 : 20 13 03 00 00 00 0c 13 24 05 
22:00:20.988 : 13 01 0c 2b 52 16 00 00 0b 
22:00:21.030 : 3b 36 00 00 0a 5a 99 06 0a 5e 
22:00:21.075 : 03 03 0a 62 97 03 0a 27 01 
22:00:21.118 : 17 04 6d 25 17 d7 2c 4c 06 43 
22:00:21.162 : 80 03 00 4c 13 22 09 07 01 
22:00:21.204 : cc 10 06 00 00 00 00 cc 20 13 
22:00:21.249 : 03 00 00 00 42 6c df 25 42 
22:00:21.292 : ec 7e ff 25 cc 01 06 44 33 03 
22:00:21.337 : 00 cc 01 13 76 06 93 00 cc 11 
22:00:21.381 : 06 00 00 00 00 cc 21 13 02 
22:00:21.424 : 00 00 00 c2 01 6c bf 2c c2 01 
22:00:21.473 : ec 7e df 2c 54 16 
```

At 22:00:20.858 we have the sequence 0c 06 which is introducing the BCD8@1 04 02 04 00 = 00040204 kWh total Energy.
Should be recognised by 1,0C06bcd8@1,Total Energy,kWh,w_total,0 but isn't...

At 22:00:21.030 we have the sequence 0a 5a which is introducing the BCD4@10 99 06 = 60.9 Degree Celsius Flow Temp. 
Should be recognised by 1,0A5Abcd4@10,Flow temp,°C,t_flow,1 but isn't...

It seems my rule doesn't fire, but I am a newbie to that.

**Reply by carsten12** [reply]
OK, Tasmota just needed some time!

The running script for Elster 96 Plus / Sharky 775 is:

```
>D
;start, define variables
wkup=1


>B
;setup sensor
->sensor53 r

>S
if upsecs%60==0 {
print read meter
=#readmeter
}
#

#readmeter

print wakeup start
;set serial protocol
ret=sml(-1 1 "2400:8N1")

;send 0x55 for 3 seconds with 8N1 (72x), 2400 baud (wakeup sequence)
for wkup 1 72 1
sml(1 1 "55555555555555555555")
next
print wakeup end
wkup=1

print wait for the meter
;wait for the meter to wake up
delay(350)

;switch serial protocol
sml(-1 1 "2400:8E1")
print request data
;set string to send to "105BFE5916" (request data)
;sml(1 1 "6804046853FE5000A116"
sml(1 1 "105BFE5916")

>M 1
+1,3,rE1,0,2400,WAERME,1
1,0C06bcd8@1,Total Energy,kWh,w_total,0
1,0C13bcd8@1000,Total volume,m³,v_total,2
1,0C2Bbcd8@1,Current power,W,p_act,0
1,0B3Bbcd6@1000,Current flow,m³/h,F_akt,3
1,0A5Abcd4@10,Flow temp,°C,t_flow,1
1,0A5Ebcd4@10,Return temp,°C,t_return,1
1,0A62bcd4@10,Temp diff,°C,t_diff,2
#
```

I will clean it up a bit and document it in Tasmota Wiki.

Thanks to everybody!

Have a nice Season,
Carsten.

**Reply by naseweiss01** [reply]
short question, my initial script for the wired MBus (see above -quoted by gemu] contains the method of counting bytes from the unique byte sequence and take the appropriate ones.
I had the issue to get wrong readings with only having two bytes as indicator - is that not going to happen with your readings?

**Reply by carsten12** [reply]
Don't understand right maybe, but Tasmota does do a relaxed pattern matching. Up to now no problems with that. What we used is a software mbus-test which is referred to in this very long German forum: https://www.mikrocontroller.net/topic/438972

This software is more aware of the M-BUS protocol. You might geht some hints there.

**Reply by naseweiss01** [reply]
Carsten,
I have tested with the relaxed pattern matching first, but I got let's say 1% wrong readings, as Tasmota randomly found matches in normal data. Therefore I modified my script to start counting the bytes from the unique sequence sequence (1,68b3b368x23uuUUUUUU ...) . As I write my readings every 2 Minutes in  a database and put them into a graph, wrong data is becoming obvious. 

I was now curious, if you have these wrong readings, too. 

In https://www.mikrocontroller.net/topic/438972 I posted my findings already, as it took me some time to get the script ready for my wired Sensostar.

**Comment by sega890** [comment]
Hi, 
I am very interested in this thread. A have also a SHARKY heating sensor but type 774. Does any have experience with that model?

Sebastian
